### ISSUE TITLE
Issue with PARAMVALUE and attribute TYPE

### ISSUE THREAD

**Issue by KSchopmeyer** [pywbem/pywbem#1241]
Submitted by Tony Garcia as email

Karl

On 04/30/2018 10:12 AM, Garcia, Tony wrote:
Hi, Andy & Karl,

I was recently trying to work with Iterators (IterEnumerateInstances and IterReferenceInstances) in pywbem 0.12.0 on a couple of VMAX systems and ran into a problem – it seems that the TYPE attribute isn’t supported in PARAMVALUE:
 
```
Traceback (most recent call last):
   File "test_iterreferences.py", line 38, in <module>
     pool = next(pool_generator, None)
   File "/usr/lib/python2.7/site-packages/pywbem/cim_operations.py", line 4080, in IterEnumerateInstances MaxObjectCount=MaxObjectCount, **extra)
   File "/usr/lib/python2.7/site-packages/pywbem/cim_operations.py", line 6187, in OpenEnumerateInstances **extra)
   File "/usr/lib/python2.7/site-packages/pywbem/cim_operations.py", line 1631, in _imethodcall tup_tree = parse_cim(tt_)
   File "/usr/lib/python2.7/site-packages/pywbem/tupleparse.py", line 334, in parse_cim child = one_child(tup_tree, ['MESSAGE', 'DECLARATION'])
   File "/usr/lib/python2.7/site-packages/pywbem/tupleparse.py", line 227, in one_child return parse_any(child)
   File "/usr/lib/python2.7/site-packages/pywbem/tupleparse.py", line 1912, in parse_any return funct_name(tup_tree)
   File "/usr/lib/python2.7/site-packages/pywbem/tupleparse.py", line 1520, in parse_message 'MULTIEXPRSP'])
   File "/usr/lib/python2.7/site-packages/pywbem/tupleparse.py", line 227, in one_child return parse_any(child)
   File "/usr/lib/python2.7/site-packages/pywbem/tupleparse.py", line 1912, in parse_any return funct_name(tup_tree)
   File "/usr/lib/python2.7/site-packages/pywbem/tupleparse.py", line 1751, in parse_simplersp child = one_child(tup_tree, ['METHODRESPONSE', 'IMETHODRESPONSE'])
   File "/usr/lib/python2.7/site-packages/pywbem/tupleparse.py", line 227, in one_child return parse_any(child)
   File "/usr/lib/python2.7/site-packages/pywbem/tupleparse.py", line 1912, in parse_any return funct_name(tup_tree)
   File "/usr/lib/python2.7/site-packages/pywbem/tupleparse.py", line 1804, in parse_imethodresponse 'PARAMVALUE'])
   File "/usr/lib/python2.7/site-packages/pywbem/tupleparse.py", line 259, in list_of_various result.append(parse_any(child))
   File "/usr/lib/python2.7/site-packages/pywbem/tupleparse.py", line 1912, in parse_any return funct_name(tup_tree)
   File "/usr/lib/python2.7/site-packages/pywbem/tupleparse.py", line 1656, in parse_paramvalue ['PARAMTYPE', 'EmbeddedObject', 'EMBEDDEDOBJECT'])
   File "/usr/lib/python2.7/site-packages/pywbem/tupleparse.py", line 185, in check_node (name(tup_tree), k))
  
pywbem.exceptions.ParseError: Element u'PARAMVALUE' has invalid attribute u'TYPE'
```

I encountered this on VMAX and VMAX3 systems. The relevant parts of the raw output that was causing the problem is:
```
<?xml version="1.0" encoding="utf-8"?>
<CIM CIMVERSION="2.0" DTDVERSION="2.0">
  <MESSAGE ID="1001" PROTOCOLVERSION="1.0">
    <SIMPLERSP>
      <IMETHODRESPONSE NAME="OpenEnumerateInstances">
 
        ...OTHER XML HERE...
 
        <PARAMVALUE NAME="EnumerationContext" TYPE="string">
          <VALUE>z3w6ett1jvdys6v73gbz----------------</VALUE>
        </PARAMVALUE>
        <PARAMVALUE NAME="EndOfSequence" TYPE="boolean">
          <VALUE>TRUE</VALUE>
        </PARAMVALUE>
      </IMETHODRESPONSE>
    </SIMPLERSP>
  </MESSAGE>
</CIM>
```

I found that simply adding TYPE as an optional attribute to the call to check_node resolved the issue:

```
def parse_paramvalue(tup_tree):
    check_node(tup_tree, 'PARAMVALUE', ['NAME'],
               ['TYPE', 'PARAMTYPE', 'EmbeddedObject', 'EMBEDDEDOBJECT'])
    child = optional_child(tup_tree,
                           ['VALUE', 'VALUE.REFERENCE', 'VALUE.ARRAY',
                            'VALUE.REFARRAY', 'CLASSNAME', 'INSTANCENAME',
                            'CLASS', 'INSTANCE', 'VALUE.NAMEDINSTANCE'])
    attrl = attrs(tup_tree)
    if 'PARAMTYPE' in attrl:
        paramtype = attrl['PARAMTYPE']
    else:
        paramtype = None
    if 'EmbeddedObject' in attrl or 'EMBEDDEDOBJECT' in attrl:
        child = parse_embeddedObject(child)
    return attrl['NAME'], paramtype, child
```

I was thinking of submitting this in a pull request and wanted to run that by you first.

**Comment by andy-maier** [comment]
We had some mail commuinication between Tony, Karl and myself, with these bottom lines:

1. CIM-XML standard perspective: Returning a TYPE attribute in a PARAMVALUE element from the server does not conform to CIM-XML. See DSP0201 2.4, section 5.3.6.5 PARAMVALUE. The respective attribute for the CIM type is named PARAMTYPE, in this case. I admit that this is an inconsistency with the name of the TYPE attribute on other elements, but that's how it is. I double checked that DSP0201 2.1 and DSP0203 2.1 (the DTD) also name the attribute PARAMTYPE, and do not have TYPE.

2. Pywbem proposal: Pywbem should implement toleration support for the TYPE attribute: If TYPE is present on a PARAMVALUE element and PARAMTYPE is not, then the value of TYPE is used where the value of PARAMTYPE would otherwise have been used. If both are present, PARAMTYPE is used and TYPE is ignored.

Actual work in pywbem:
- tupleparse.py: Tolerate TYPE and handle like PARAMTYPE is handled. Add a comment in the code as to why this non-standard behavior is supported.
- test_tupleparse.py: Add testcases for TYPE instead of PARAMTYPE, and TYPE with PARAMTYPE.
- testclient/invokemethod.yaml: Add testcase for TYPE instead of PARAMTYPE, and TYPE with PARAMTYPE.

**Discussion point:** Should the presence of TYPE on PARAMVALUE cause a Python `UserWarning` to be issued?

**Comment by andy-maier** [comment]
Tomy wants to bring forward a PR that implements the proposal.

**Comment by andy-maier** [comment]
My take on the discussion point is that we should not issue a UserWarning, for consistency with how we handle the non-standard `EMBEDDEDOBJECT` attribute pywbem tolerates (returned by older versions of OpenPegasus).

**Comment by KSchopmeyer** [comment]
So the only change is that if PARAMTYPE is not provided, pywbem tries to use non-standard TYPE.

One question: How does this code get through the SNIA SMI CTP tests? Did they implement an exception?

**Comment by andy-maier** [comment]
Is usage of the pull operations part of those tests?

**Comment by KSchopmeyer** [comment]
Tony.  The pr you created https://github.com/pywbem/pywbem/pull/1274  is up but has some issues.   The question today is do you want to finish it up or pass it to us to finish and commit??

Would you get back to us on this???

**Comment by andy-maier** [comment]
Fixed in PR #1274.
